cognitive/src/main/scala/com/microsoft/ml/spark/cognitive/FormRecognizer.scala

```diff
@@ -0,0 +1,634 @@
+    val fromRow = AnalyzeIDDocumentsResponse.makeFromRowConverter
+    def extractText(lines: Array[ReadLine]): String = {
+      lines.map(_.text).mkString(" ")
+    }
+    new UDFTransformer()
+      .setUDF(UDFUtils.oldUdf(
+        { r: Row =>
+          Option(r).map(fromRow).map(
+            _.analyzeResult.readResults.map(_.lines.map(extractText).mkString("")).mkString(" ")).mkString("")
+        },
+        StringType))
+      .setInputCol(inputCol)
+      .setOutputCol(outputCol)
+  }
+
+  def flattenDocumentResults(inputCol: String, outputCol: String): UDFTransformer = {
+    val fromRow = AnalyzeIDDocumentsResponse.makeFromRowConverter
+    def extractFields(documentResults: Seq[DocumentResult]): String = {
+      documentResults.map(_.fields).mkString("\n")
+    }
+    new UDFTransformer()
+      .setUDF(UDFUtils.oldUdf(
+        { r: Row =>
+          Option(r).map(fromRow).map(
+            _.analyzeResult.documentResults.map(extractFields).mkString("")).mkString("")
+        },
+        StringType))
+      .setInputCol(inputCol)
+      .setOutputCol(outputCol)
+  }
+}
+
+class AnalyzeIDDocuments(override val uid: String) extends FormRecognizerBase(uid) with BasicLogging {
+  logClass()
+
+  def this() = this(Identifiable.randomUID("AnalyzeIDDocuments"))
+
+  def setLocation(v: String): this.type =
+    setUrl(s"https://$v.api.cognitive.microsoft.com/formrecognizer/v2.1/prebuilt/idDocument/analyze")
+
+  val includeTextDetails = new ServiceParam[Boolean](this, "includeTextDetails",
+    "Include text lines and element references in the result.", isURLParam = true)
+
+  def  setIncludeTextDetails(v: Boolean): this.type = setScalarParam(includeTextDetails, v)
+
+  setDefault(includeTextDetails -> Left(false))
+
+  val pages = new ServiceParam[String](this, "pages", "The page selection for multi-page PDF" +
+    " and TIFF documents, to extract Receipt information from individual pages and a range of pages (like page 2," +
+    " and pages 5-7) by entering the page numbers and ranges separated by commas (e.g. '2, 5-7'). If not set," +
+    " all pages will be processed.", isURLParam = true)
+
+  def setPages(v: String): this.type = setScalarParam(pages, v)
+
+  def setPagesCol(v: String): this.type = setVectorParam(pages, v)
+
+  override protected def responseDataType: DataType = AnalyzeIDDocumentsResponse.schema
+
+}
+
+object TrainCustomModel extends ComplexParamsReadable[TrainCustomModel] {
+  def getModelId(inputCol: String, outputCol: String): UDFTransformer = {
+    val fromRow = TrainCustomModelResponse.makeFromRowConverter
+    new UDFTransformer()
+      .setUDF(UDFUtils.oldUdf(
+        { r: Row =>
+          Option(r).map(fromRow).map(_.modelInfo.modelId).mkString("")
+        },
+        StringType))
+      .setInputCol(inputCol)
+      .setOutputCol(outputCol)
+  }
+}
+
+class TrainCustomModel(override val uid: String) extends CognitiveServicesBaseNoHandler(uid)
```

Comment:
I see this is a transformer, but training to me invokes the idea of a SparkML estimator. Though this might be best suited for a different class we should discuss how a natural sparkML design of a custom AI solution might look. My thought was that we could have something that is fit to a dataframe of forms and return the corresponding AnalyzeCustomModel transform with the modelID set with a literal value (so that it applies the same custom model to all columns)